assets/src/domain/eventEditor/ui/ticketAssignmentsManager/utils.ts

```diff
@@ -36,6 +37,76 @@ export const prepareEntitiesForUpdate = <Entity extends TAMRelationEntity>({
 	}, Object.entries(newEntities));
 };
 
+type TicketsWithQuantityArgs = {
+	ticketsToUpdate: EntitiesToUpdate;
+	existingData: TAMRelationalData;
+	allDates: Array<Datetime>;
+	allTickets: Array<Ticket>;
+};
+
+/**
+ * Returns the tickets to quantity map of the tickets that need update
+ */
+export const ticketsWithNewQuantity = ({
+	allDates,
+	allTickets,
+	existingData,
+	ticketsToUpdate,
+}: TicketsWithQuantityArgs): AnyObject<number> => {
+	// create a map of date ids to capacities
+	const dateIdToCapacityMap: AnyObject<number> = allDates.reduce(
+		(acc, date) => assocPath([date.id], date.capacity, acc),
+		{}
+	);
+	// create a map of ticket ids to quantities
+	const ticketIdToQuantityMap: AnyObject<number> = allTickets.reduce(
+		(acc, ticket) => assocPath([ticket.id], ticket.quantity, acc),
+		{}
+	);
+	/**
+	 * This becomes an object with key as ticket id and value as array of new date ids
+	 * {
+	 *     [ticket.id]: quantity
+	 * }
+	 */
+	return ticketsToUpdate.reduce<AnyObject<number>>((acc, [ticketId, possibleRelation]) => {
+		// These are the related date ids before TAM was launched
+		const existingRelatedDateIds = existingData?.tickets?.[ticketId]?.datetimes || [];
+		// These are the related date ids that have been assigned in TAM
+		// these contain old and newly assigned date ids
+		const allNewRelatedDateIds: Array<EntityId> = possibleRelation?.datetimes || [];
+
+		// lets filter out the existing ones
+		const newOnlyRelatedDateIds = allNewRelatedDateIds.filter(
+			// it's new only if it was not in existing relations
+			(dateId) => !existingRelatedDateIds.includes(dateId)
+		);
+
+		// lets convert the array of ids to array of corresponding capacities
+		const newDateCapacities = newOnlyRelatedDateIds
+			// get capacity from the above map and parse it as infinity
+			.map((dateId) => parseInfinity(dateIdToCapacityMap?.[dateId], Infinity))
+			// we don't need to update ticket quantity if date capacity is infinite
+			.filter((capacity) => !isInfinite(capacity));
+
+		// we need to set the ticket quantity to the minimum of all the capacities
+		const minimumCapacity = Math.min(...newDateCapacities); // it will be Infinity for empty array
+
+		// Make sure that the non negative ticket quantity value is compared with
+		// a non negative datetime capacity value in Math.min()
+		const nonNegativeTicketQuantity = parseInfinity(ticketIdToQuantityMap?.[ticketId], Infinity);
+
+		// if capacity is infinite or it's more than ticket quantity
+		if (isInfinite(minimumCapacity) || minimumCapacity > nonNegativeTicketQuantity) {
+			// no need to update the ticket quantity
+			return acc;
+		}
+
+		// set the quantity to minimum capacity
+		return assocPath([ticketId], minimumCapacity, acc);
+	}, {});
+};
```

Comment:
beautiful job making that complex logic easy to read and understand 💯

Reply:
agree, this is well-written code;
this speaks to me of high level of caring for reviewers and ease of future development/ refactoring